[New Blog Post] Orchestrating Events with Knative and Kogito (#3083)
knative/docs: blog/articles/event-drive-app-knative-eventing-kogito.md

---
Title: 'Orchestrating Events with Knative and Kogito'
Author: Ricardo Zanini and Tihomir Surdilovic
Author handle: https://twitter.com/zaninirica, https://twitter.com/tsurdilo

> Date: ''

Reviewer: Note: Remember to add a date when we merge :slightly_smiling_face:

Author: Which date should I add? Maybe today?

Description: A brief introduction of Knative Eventing integration with Kogito implementation of the Serverless Workflow specification
Folder with media files: 'images/kogito'
Blog URL: ''
Labels: Demo
Reviewers: ''

---
| Reviewer  | Date | Approval |
| ------------- | ------------- | ------------- |
| <!-- Your Github handle here -->  |   |:+1:, :monocle_face:, :-1:  |

> | <!-- Your Github handle here --> |   |  |

Reviewer: Note: Add me when merging :slightly_smiling_face: anyone else who needs to sign off?

[Kogito](https://kogito.kie.org/) is a platform for the development of cloud-native business automation applications. Kogito is designed targeting cloud-native architectures, and it comes with a series of features to make it easy for architects and developers to create business applications.

> Kogito implements the [CNCF Serverless Workflow Project](https://github.com/serverlessworkflow/specification), which is a specification for defining workflow models that orchestrate event-driven, serverless applications. It focuses on defining a vendor-neutral, platform-independent, and declarative workflow model for orchestrating services that can be used across multiple cloud/container platforms. To this date, the Serverless Workflow specification is a CNCF sandbox project.

Reviewer: ```suggestion
Kogito implements the [CNCF Serverless Workflow Project](https://github.com/serverlessworkflow/specification), which is a specification for defining workflow models that orchestrate event-driven, serverless applications. It focuses on defining a vendor-neutral, platform-independent, and declarative workflow model for orchestrating services that can be used across multiple cloud and container platforms. To date, the Serverless Workflow specification is a CNCF sandbox project.
```

As part of the Serverless Workflow implementation, Kogito offers a Kubernetes Operator to deploy these workflows with Knative. The goal is to make it as simple as possible to deploy and manage user-defined workflows in  cloud environments. Knative Eventing plays a very important role in this scenario by providing the underlying infrastructure for event-driven architectures.

### Kogito Serverless Workflow

> To demonstrate how the Kogito workflow implementation works on event-driven architectures, we will use the [Patient Onboarding example](https://github.com/kiegroup/kogito-examples/tree/master/serverless-workflow-functions-events-quarkus). In this example, we simulate a workflow used in hospitals to onboard new patients and assign them to the correct doctor.

Reviewer: ```suggestion
To demonstrate how the Kogito workflow implementation works on event-driven architectures, we will use the [patient onboarding example](https://github.com/kiegroup/kogito-examples/tree/master/serverless-workflow-functions-events-quarkus). In this example, we simulate a workflow used in hospitals to onboard new patients and assign them to the correct doctor.
```

The following image taken from the [specification examples page](https://github.com/serverlessworkflow/specification/blob/master/examples/examples.md#New-Patient-Onboarding) illustrates this workflow:

> ![Patient Onboarding workflow representation](https://raw.githubusercontent.com/serverlessworkflow/specification/master/media/examples/example-patientonboarding.png)*Patient Onboarding workflow representation*

Reviewer: ```suggestion
![Patient onboarding workflow representation](https://raw.githubusercontent.com/serverlessworkflow/specification/master/media/examples/example-patientonboarding.png)*Patient Onboarding workflow representation*
```

You can just make the changes yourself instead of accepting the suggestion if it causes you to need to rebase etc, but I'll add them as suggestions anyway so the change is highlighted.

> The workflow will start after receiving a [Cloud Event](https://github.com/cloudevents/spec) with patient information, and then will call three functions in a sequence which: (1) store the patient information, (2) assign to a doctor based on their symptoms, and (3) schedule an appointment with the assigned doctor.

Reviewer: ```suggestion
The workflow starts after receiving a [CloudEvent](https://github.com/cloudevents/spec) object that contains patient information. Three functions are then called by the spec in a sequence which: (1) stores the patient information, (2) assigns the patient to a doctor based on their symptoms, and (3) schedules an appointment with the assigned doctor for that patient.
```

> This is how the Patient Onboarding workflow looks like based on the specification:

Reviewer: ```suggestion
Here is an example of the patient onboarding workflow YAML, based on this specification:
```

```yaml
id: onboarding
version: '1.0'
name: Patient Onboarding Workflow
states:
- name: Onboard
  type: event
  start:
    kind: default
  onEvents:
  - eventRefs:
    - NewPatientEvent
    actionMode: sequential
    actions:
    - functionRef:
        refName: StoreNewPatient
    - functionRef:
        refName: AssignPatientToDoctor
    - functionRef:
        refName: SchedulePatientAppointment
  end:
    kind: default
events:
- name: NewPatientEvent
  type: new.patient.events
  source: "/hospital/entry"
functions:
- name: StoreNewPatient
  operation: classpath:openapi.json#storeNewPatient
- name: AssignPatientToDoctor
  resource: classpath:openapi.json#assignPatientToDoctor
- name: SchedulePatientAppointment
  resource: classpath:openapi.json#schedulePatientAppointment
```

> The above workflow definition is based on three parts. The first part is the “states” array which includes the workflow control-flow logic, in our case the “event” state which receives the new patient event and performs our three actions (service invocations).

Reviewer: ```suggestion
This specification is based on three parts. The first part is the `states` array, which includes the workflow control flow logic. In this case, the state is an `event`, which receives the new patient cloud event, and performs the three functions mentioned previously.
```

Author: Hi @abrennan89 many thanks for your review! :smile: 

This paragraph is about the workflow definition listed above. I think "This specification" won't reflect that, so the reader wouldn't associate the workflow and the description that follows it.

Reviewer 2: Would it be more clear to start that sentence with
"The above workflow definition contains three mayor parts ...."
The entire workflow definition complies with the specification. Just trying to figure out what you think would be best wording for it.

Reviewer: That's fine - could you please make the changes to `states` etc though, since they're literal fields so it seems like they should be formatted like `this` instead of "this"? I don't mind what way you choose to word this part though, and thanks for clarifying :)

> The second part is our event definitions (the “events” array), which are based on the CloudEvents specification, so there is a clear 1-1 mapping between how the events are defined in the workflows, and how they are represented with the CloudEvents format.

Reviewer: ```suggestion
The second part of the spec is the `events` array, which contains event definitions that are based on the CloudEvent specification. There is a 1:1 mapping between how the events are defined in the workflows, and how they are represented with the CloudEvent format.
```

Reviewer 2: Same here, I think this paragraph is specific to the actual workflow definition shown, and not the specification itself. 
From the spec perspective valid workflows that don't consume events would not define the "events" array", so this is specifically describing the example workflow in the doc.

> The third part, the “functions” array, is the function definitions which give more information to our runtime about how to execute the needed services.

Reviewer: ```suggestion
The third part, the `functions` array, contains the function definitions that give more information to the runtime about how to execute the required services.
```

Serverless Workflow specification is standards-based and leverages the OpenAPI specification to define details about restful executions of services. [In this gist](https://gist.github.com/ricardozanini/5ec66b4ddfbcf8ab40747b28b5f86333), you can see the OpenAPI file referenced by the functions in the above example.

You can reference the [Serverless Workflow specification](https://github.com/serverlessworkflow/specification/blob/master/specification.md) for information about all language constructs that the specification provides.

> Note that the specification allows for both JSON and YAML workflow formats, and we used the YAML one in this example, but one based on JSON is considered equivalent and also parsable by the runtime.

Reviewer: ```suggestion
Note that the specification allows for both JSON and YAML workflow formats. This example uses YAML, however JSON is considered equivalent and is also parsable by the runtime.
```

During compilation, the Kogito runtime will parse this YAML file and will generate Java code that represents this workflow definition. The generated code is based on the Quarkus framework. The outcome is an OpenAPI standard REST service that can be deployed anywhere in your architecture.

![Kogito Runtime workflow parse process](images/kogito/kogito-process-workflow-file.png)*Kogito Runtime parsing flow*

> In this example, we also added the [Knative Kogito Eventing plugin](https://docs.jboss.org/kogito/release/latest/html_single/#con-knative-eventing_kogito-developing-process-services)  to the project, which means that it can accept Cloud Events through HTTP on the root path, for example:

Reviewer: ```suggestion
In this example, we have also added the [Knative Kogito Eventing plugin](https://docs.jboss.org/kogito/release/latest/html_single/#con-knative-eventing_kogito-developing-process-services) to the project, which means that it can accept CloudEvent objects through HTTP on the root path. For example:
```

```bash
$ curl -X POST \
      -H "content-type: application/json"  \
      -H "ce-specversion: 1.0"  \
      -H "ce-source: /hospital/entry"  \
      -H "ce-type: new.patients.events"  \
      -H "ce-id: 12346"  \
      -d "{ \"name\": \"Mick\", \"dateOfBirth\": \"1983-08-15\", \"symptoms\":[\"seizures\"]}" \
http://localhost:8080
```

> When sending this request to the service, it will start a new instance of the workflow and perform all the operations defined by it.

Reviewer: ```suggestion
When sending this request to the service, a new instance of the workflow starts, and performs all of the operations defined in the spec.
```

There's much more included in this example. For comprehensive instructions on how to deploy and build it, please see the example page on the [Github repository](https://github.com/kiegroup/kogito-examples/tree/master/serverless-workflow-functions-events-quarkus).

### Knative Eventing Integration

> Based on this generated service, [an image can be built](https://docs.jboss.org/kogito/release/latest/html_single/#proc-kogito-deploying-on-kubernetes_kogito-deploying-on-openshift) to be deployed with the [Kogito Operator](https://docs.jboss.org/kogito/release/latest/html_single/#con-kogito-operator-and-cli_kogito-deploying-on-openshift) on a Kubernetes cluster with [Knative Eventing installed](https://knative.dev/docs/eventing/). The operator will create all the necessary Knative resources to have this service configured and listening to a [Broker](https://knative.dev/docs/eventing/broker/):

Reviewer: ```suggestion
Based on this generated service, you can build [an image](https://docs.jboss.org/kogito/release/latest/html_single/#proc-kogito-deploying-on-kubernetes_kogito-deploying-on-openshift) to be deployed with the [Kogito Operator](https://docs.jboss.org/kogito/release/latest/html_single/#con-kogito-operator-and-cli_kogito-deploying-on-openshift) on a Kubernetes cluster with [Knative Eventing](https://knative.dev/docs/eventing/) installed. The Operator will create all the necessary Knative resources to configure this service and subscribe it to the Knative [broker](https://knative.dev/docs/eventing/broker/):
```

![Knative and Kogito integration](images/kogito/knative-eventing-kogito-operator.png)*Knative Eventing and Kogito Operator integration*

> The Kogito Operator will create a [Knative Trigger](https://knative.dev/docs/eventing/triggers/) resource that will link the service and the Broker together. In this example, it will filter events of type `new.patients.events`. Meaning that every time a new event of this type comes to the Broker, it will be redirected to the Kogito service.

Reviewer: ```suggestion
The Kogito Operator creates a Knative [Trigger](https://knative.dev/docs/eventing/triggers/) resource, that links the service and the broker together. In this example, it will filter events of type `new.patients.events`. This means that every time a new event of this type comes to the broker, it will be redirected to the Kogito service.
```

> The same concept also applies to [events produced by the workflow engine](https://github.com/serverlessworkflow/specification/blob/0.5.x/specification.md#event-definition). In this case, the operator will create a [Knative SinkBinding](https://knative.dev/docs/eventing/samples/sinkbinding/) resource and will bind it to a given Broker. Each time an event is produced by the service, a Cloud Event representing it will be transmitted to the Broker.

Reviewer: ```suggestion
The same concept also applies to [events produced by the workflow engine](https://github.com/serverlessworkflow/specification/blob/0.5.x/specification.md#event-definition). In this case, the operator will create a Knative [SinkBinding](https://knative.dev/docs/eventing/samples/sinkbinding/) resource, and will bind it to the Knative broker. Each time an event is produced by the service, a CloudEvent representing it will be sent to the broker.
```

![Kogito Runtime workflow parse process](images/kogito/kogito-process-workflow-file.png)*Kogito Runtime parsing flow*

In this example, we have also added the [Knative Kogito Eventing plugin](https://docs.jboss.org/kogito/release/latest/html_single/#con-knative-eventing_kogito-developing-process-services) to the project, which means that it can accept CloudEvent objects through HTTP on the root path. For example:

```bash
$ curl -X POST \
      -H "content-type: application/json"  \
      -H "ce-specversion: 1.0"  \
      -H "ce-source: /hospital/entry"  \
      -H "ce-type: new.patients.events"  \
      -H "ce-id: 12346"  \
      -d "{ \"name\": \"Mick\", \"dateOfBirth\": \"1983-08-15\", \"symptoms\":[\"seizures\"]}" \
http://localhost:8080
```

When sending this request to the service, a new instance of the workflow starts, and performs all of the operations defined in the spec.

There's much more included in this example. For comprehensive instructions on how to deploy and build it, please see the example page on the [Github repository](https://github.com/kiegroup/kogito-examples/tree/master/serverless-workflow-functions-events-quarkus).

### Knative Eventing Integration

> Based on this generated service, you can build [an image](https://docs.jboss.org/kogito/release/latest/html_single/#proc-kogito-deploying-on-kubernetes_kogito-deploying-on-openshift) to be deployed with the [Kogito Operator](https://docs.jboss.org/kogito/release/latest/html_single/#con-kogito-operator-and-cli_kogito-deploying-on-openshift) on a Kubernetes cluster with [Knative Eventing](https://knative.dev/docs/eventing/) installed. The Operator will create all the necessary Knative resources to configure this service and subscribe it to the Knative [broker](https://knative.dev/docs/eventing/broker/):

Reviewer 3: trailing ':'

> ![Knative and Kogito integration](images/kogito/knative-eventing-kogito-operator.png)*Knative Eventing and Kogito Operator integration*

Reviewer 3: This line didn't seem to render properly for me in the preview

Author: hmmm we need a new line here.

The Kogito Operator creates a Knative [Trigger](https://knative.dev/docs/eventing/triggers/) resource, that links the service and the broker together. In this example, it will filter events of type `new.patients.events`. This means that every time a new event of this type comes to the broker, it will be redirected to the Kogito service.

> The same concept also applies to [events produced by the workflow engine](https://github.com/serverlessworkflow/specification/blob/0.5.x/specification.md#event-definition). In this case, the operator will create a Knative [SinkBinding](https://knative.dev/docs/eventing/samples/sinkbinding/) resource, and will bind it to the Knative broker. Each time an event is produced by the service, a CloudEvent representing it will be sent to the broker.

Reviewer 3: could we be explicit in what the `sinkbinding` subject is here?  If we're using yaml for this example, would it be possible to link to the yaml resource?

Author: Hi @lberk ! Thanks for your review. The `sinkbinding` here is created automatically by the Kogito Operator once a workflow is deployed, so we don't have the yaml source (I could create an example thou). See: 
https://docs.jboss.org/kogito/release/latest/html_single/#con-knative-eventing_kogito-developing-process-services

I think we can link this doc here?

In this particular example, we only have the `trigger` to be created since we listening to the new patient event.

Reviewer 3: Hi! Fair enough, I think I was just trying to say 'I'm not sure what exactly the subject is' from the link you provided I think this diagram summarizes it (and if we could figure out how to communicate this somehow in the blog post, it'd be great!) https://docs.jboss.org/kogito/release/latest/html_single/kogito/bpmn/kogito-knative-impl-producing-event.png

Author: @lberk I pushed an update to the article incorporating this image and introducing it to this context, so readers can have a better view of the concept. I don't want to extend too much since this is an introductory article about this integration. I intend to keep writing more about these concepts in the future using this article as a reference. wdyt?
